keps/2724-topology-aware-schedling/README.md

```diff
@@ -463,28 +471,72 @@ at each topology level to the specific subset of nodes.
 type PodSetAssignment struct {
   ...
 
-  // TopologyAssignment indicates the resources assigned per topology level
+  // topologyAssignment indicates the topology assignment divided into
+  // topology domains corresponding to the lowest level of the topology.
+  // The assignment specifies the number of Pods to be scheduled per topology
+  // domain and specifies the node selectors for each topology domain, in the
+  // following way: the node selector keys are specified by the levels field
+  // (same for all domains), and the corresponding node selector value is
+  // specified by the domains.values subfield.
+  //
+  // Example:
+  //
+  // topologyAssignment:
+  //   levels:
+  //   - cloud.provider.com/topology-block
+  //   - cloud.provider.com/topology-rack
+  //   domains:
+  //   - values: [block-1, rack-1]
+  //     count: 4
+  //   - values: [block-1, rack-2]
+  //     count: 2
+  //
+  // Here:
+  // - 4 Pods are to be scheduled on nodes matching the node selector:
+  //   cloud.provider.com/topology-block: block-1
+  //   cloud.provider.com/topology-rack: rack-1
+  // - 2 Pods are to be scheduled on nodes matching the node selector:
+  //   cloud.provider.com/topology-block: block-1
+  //   cloud.provider.com/topology-rack: rack-2
+  //
   // +optional
   TopologyAssignment *TopologyAssignment `json:"topologyAssignment,omitempty"`
 }
 
 type TopologyAssignment struct {
-  // Groups contains the list of assignments split into groups corresponding
-  // to the same topology domain at the lowest level of the hierarchy.
+  // levels is an ordered list of keys denoting the levels of the assigned
+  // topology (i.e. node label keys), from the highest to the lowest level of
+  // the topology.
+  //
   // +required
   // +listType=atomic
   // +kubebuilder:validation:MinItems=1
-  Groups []TopologyAssignmentGroup `json:"groups"`
+  // +kubebuilder:validation:MaxItems=8
+  Levels []string `json:"levels"`
```

Comment:
Which objectives does this field "Storing desired state" or "temporary cache for the computing by topologyUngator"?
I guess that this is for the "temporary cache for the computing by topologyUngator".

In that case, could we put this information in the internal cache?

Reply:
It is desired state to be reached by the Pods. We cannot use internal state,  because then it would be lost on Kueue restart.

Reply:
Does that mean that the topologyUngator can not obtain levels from Topology CR when the Kueue restarts?

Reply:
Oh, I see what you mean. Similar question was asked in this [comment](https://github.com/kubernetes-sigs/kueue/pull/3237#discussion_r1801305886). 

It makes the implementation of TopologyUngater simpler (as it would require lookups via RF API Topology API, and dealing with changes to the object), at the very small cost in terms of object size as the number of levels is quite limited.

Also, I think it is conceptually consistent with what we do [here](https://github.com/kubernetes-sigs/kueue/blob/d9db44460c0b61532b3ca03c7f98ef935dd848ff/apis/kueue/v1beta1/workload_types.go#L98-L99) as we store the mapping between resources and flavors, rather than reading resources from the CQ API object.

Reply:
> as it would require lookups via RF API Topology API, and dealing with changes to the object

I think that this could be resolved by the dedicated internal cache to store the relarationships between RF and Topology.

> at the very small cost in terms of object size as the number of levels is quite limited.

Yes, I agree with you. My doubt was that the field is just cache since we can obtain from another objects.

> Also, I think it is conceptually consistent with what we do [here](https://github.com/kubernetes-sigs/kueue/blob/d9db44460c0b61532b3ca03c7f98ef935dd848ff/apis/kueue/v1beta1/workload_types.go#L98-L99) as we store the mapping between resources and flavors, rather than reading resources from the CQ API object.

Uhm, I see. Indeed the levels (new field) and flavors (existing field) field is stored in the Workload status, and it seems that reduce the multiple API calls (RF -> Topology).

When we graduate API version to v1beta2, we may want to restructure the Workload status field so that we can store the actual assigned flavor and topology information in a same level field.

Anyway, I'm ok with adding the levels here. Thanks.

Reply:
> I think that this could be resolved by the dedicated internal cache to store the relarationships between RF and Topology.

at the very small cost in te

Reply:
>When we graduate API version to v1beta2, we may want to restructure the Workload status field so that we can 
> store the actual assigned flavor and topology information in a same level field.
> Sure, I'm ok to add it to the KEP to re-evaluate this decision.

let me know if you want that to be added, though I think it is unlikely to be dropped. For example, if we read from cache the keys could be changed in the meanwhile, and then the domain assignement would be corrupted. If we know the keys at the moment of assignment we could compare them with the ones in the cache and evict the workload.

Reply:
> sure, it looks possible, I get the point, but I think it would complicate the implementation to maintain the cache, rather than reading from the object.

Yeah, that's true. I think that this is a trade-off between implementation costs and API maintenance costs since the dedicated cache could prevent exposing API and easily break the structure, but it has implementation costs.

Anyway, I do not claim a dedicated cache now.

> For example, if we read from cache the keys could be changed in the meanwhile, and then the domain assignement would be corrupted.

This could be preventented by the event based cache updating same as today cache package (/pkg/cache).
I'm wondering if we can put the dedicated cache mechanism instead of the status field on the alternative section.
But this is nonblocking of this PR merging.

Reply:
I can update the alternative section in a follow up

Reply:
Yeah, sure.